### PR TITLE
Auto relativize PDF file paths when selecting "Search for unlinked local files"

### PR DESCRIPTION
**Pull request by u7676493**
Fixes [koppor#549](https://github.com/koppor/jabref/issues/549)

It updates the functionality of the "Search for unlinked local files" import option - it now automatically relativizes the file paths for PDF files.

### List of changes

- Modified `ImportHandler` to use different arguments for `importPDFContent()`, accounting for the changes made to `ExternalFilesContentImporter`
  `ImportHandler` starts the execution of the logic of importing a PDF file.
- Modified `ExternalFilesContentImporter` to include imports for `BibDatabaseContext` and `FilePreferences` classes as well as different arguments for `importPDFContent()` and `importDatabase()`, both related to the changes made to `PdfMergeMetadataImporter`
- Modified `PdfMergeMetadataImporter` to add an alternate `importDatabase()` function with different arguments, using the `BibDatabaseContext.getFileDirectories(FilePreferences)` function to get all file directories and relativize the file path using the `FileUtil.relativize(Path, List<Path>)` function, before passing it into the default override of the `importDatabase(Path)` function from the `Importer` class.
  `PdfMergeMetadataImporter` handles all of the actual logic relating to importing a PDF file.
  The new `importDatabase()` function has three arguments: `(Path file, BibDatabaseContext context, FilePreferences preferences)`
- Added a WIP test in `PdfMergeMetadataImporter`

Please provide me with any feedback you can! I haven't included any changes in documentation yet because this is still a WIP.

### Mandatory checks

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java [hunk 1]
```diff
@@ -118,7 +118,7 @@ public List<ImportFilesResultItemViewModel> call() {
 
                     try {
                         if (FileUtil.isPDFFile(file)) {
-                            var pdfImporterResult = contentImporter.importPDFContent(file);
+                            var pdfImporterResult = contentImporter.importPDFContent(file, bibDatabaseContext, filePreferences);
```
Comment (subhramit) [inline src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:121]: Can you change this to the strict returned type here?

File: src/main/java/org/jabref/logic/importer/fileformat/PdfMergeMetadataImporter.java [hunk 2]
```diff
@@ -136,6 +136,29 @@ public ParserResult importDatabase(Path filePath) throws IOException {
...
 
+    /**
+     * A modified version of {@link PdfMergeMetadataImporter#importDatabase(Path)}, but it
+     * relativizes the {@code filePath} if there are working directories before parsing it
+     * into {@link PdfMergeMetadataImporter#importDatabase}
+     * (Otherwise no path modification happens).
+     *
+     * @param filePath    The unrelativized {@code filePath}.
+     * @param context     Everything related to the BIB file
+     * @param preferences Preferences for linked files
```
Comment (subhramit) [inline src/main/java/org/jabref/logic/importer/fileformat/PdfMergeMetadataImporter.java:147]: 146 - To be precise, context is the metadata for the library. However, it's a common JabRef parameter, so can remove it.
147 - Can remove as well, as the type makes the description redundant.

File: src/test/java/org/jabref/logic/importer/fileformat/PdfMergeMetadataImporterTest.java [hunk 3]
```diff
@@ -84,4 +86,28 @@ void importWorksAsExpected() throws Exception {
...
+        // Initialize file and working directory
+        Path file = Path.of(PdfMergeMetadataImporter.class.getResource("/pdfs/minimal.pdf").toURI());
+        Path directory = Path.of(PdfMergeMetadataImporter.class.getResource("/pdfs/").toURI());
+        preferences.setWorkingDirectory(directory);
+
+        // Set up results
+        List<BibEntry> result = importer.importDatabase(file, database, preferences).getDatabase().getEntries();
+
+        // Set up expected value
+        BibEntry expected = new BibEntry(StandardEntryType.InProceedings)
```
Comment (subhramit) [inline src/test/java/org/jabref/logic/importer/fileformat/PdfMergeMetadataImporterTest.java:105]: Remove these two comments (redundant and standard).

**Review by subhramit — CHANGES_REQUESTED** [review]

**Comment by u7676493** [comment]
I have just added a commit that follows up on all of the requested changes so far.

File: src/main/java/org/jabref/logic/externalfiles/ExternalFilesContentImporter.java [hunk 4]
```diff
@@ -17,9 +19,9 @@ public ExternalFilesContentImporter(ImportFormatPreferences importFormatPreferen
         this.importFormatPreferences = importFormatPreferences;
     }
 
-    public ParserResult importPDFContent(Path file) {
+    public ParserResult importPDFContent(Path file, BibDatabaseContext context, FilePreferences preferences) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/externalfiles/ExternalFilesContentImporter.java:22]: Name the `preferences` variable `filePreferences` - also in the method `importDatabase`

**Review by koppor — CHANGES_REQUESTED** [review]
Tried it out. Works.

Small naming comment - otherwise code looks good. No need for additional tests. - Code is pretty similar with https://github.com/JabRef/jabref/pull/10582 - which we also reviewed back then. - The PR was abandoned, thank you for porting it to the most recent version of JabRef.

Please add CHANGELOG.md entry and then it should be good to go!

**Review by koppor — APPROVED** [review]

**Comment by koppor** [comment]
@u7676493 I already added it to the merge queue. You updated. Do you want to do any more changes?

File: src/main/java/org/jabref/logic/importer/fileformat/PdfMergeMetadataImporter.java [hunk 5]
```diff
@@ -136,6 +136,27 @@ public ParserResult importDatabase(Path filePath) throws IOException {
...
+     * @param filePath    The unrelativized {@code filePath}.
+     */
+    public ParserResult importDatabase(Path filePath, BibDatabaseContext context, FilePreferences filePreferences) throws IOException {
+        Objects.requireNonNull(context);
+        Objects.requireNonNull(filePreferences);
+
+        List<Path> directories = context.getFileDirectories(filePreferences);
+
+        if (!directories.isEmpty()) {
+            filePath = FileUtil.relativize(filePath, directories);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/PdfMergeMetadataImporter.java:154]: Doesn't that method also have a check for emptyness? Then, this call can be moved out of the if and the jf be removed.

Comment (u7676493) [reply src/main/java/org/jabref/logic/importer/fileformat/PdfMergeMetadataImporter.java]: I have just added a commit that does that

**Review by subhramit — APPROVED** [review]
Looks good. Adding back to merge queue.

**Review by subhramit — APPROVED** [review]

**Pull request merged**